spark/src/test/scala-spark-master/org/apache/spark/sql/delta/DeltaVariantShreddingSuite.scala

```diff
@@ -0,0 +1,174 @@
+      assert(!deltaLog.unsafeVolatileSnapshot.protocol
+        .isFeatureSupported(VariantShreddingPreviewTableFeature),
+        s"Table tbl contains ShreddedVariantTableFeature descriptor when its not supposed to"
+      )
+      sql(s"ALTER TABLE tbl " +
+        s"SET TBLPROPERTIES('${DeltaConfigs.ENABLE_VARIANT_SHREDDING.key}' = 'true')")
+      assert(getProtocolForTable("tbl")
+        .readerAndWriterFeatures.contains(VariantShreddingPreviewTableFeature))
+    }
+    withTable("tbl") {
+      sql(s"CREATE TABLE tbl(s STRING, i INTEGER) USING DELTA " +
+        s"TBLPROPERTIES('${DeltaConfigs.ENABLE_VARIANT_SHREDDING.key}' = 'true')")
+      assert(getProtocolForTable("tbl")
+        .readerAndWriterFeatures.contains(VariantShreddingPreviewTableFeature))
+    }
+    assert(DeltaConfigs.ENABLE_VARIANT_SHREDDING.key == "delta.enableVariantShredding")
+  }
+
+  test("Spark can read shredded table containing the shredding table feature") {
+    withTable("tbl") {
+      withTempDir { dir =>
+        val schema = "a int, b string, c decimal(15, 1)"
+        val df = spark.sql(
+          """
+            | select id i, case
+            | when id = 0 then parse_json('{"a": 1, "b": "2", "c": 3.3, "d": 4.4}')
+            | when id = 1 then parse_json('{"a": [1,2,3], "b": "hello", "c": {"x": 0}}')
+            | when id = 2 then parse_json('{"A": 1, "c": 1.23}')
+            | end v from range(0, 3, 1, 1)
+            |""".stripMargin)
+
+        sql("CREATE TABLE tbl (i long, v variant) USING DELTA " +
+          s"TBLPROPERTIES ('${DeltaConfigs.ENABLE_VARIANT_SHREDDING.key}' = 'true') " +
+          s"LOCATION '${dir.getAbsolutePath}'")
+        assert(getProtocolForTable("tbl")
+          .readerAndWriterFeatures.contains(VariantShreddingPreviewTableFeature))
+        withSQLConf(SQLConf.VARIANT_WRITE_SHREDDING_ENABLED.key -> true.toString,
+          SQLConf.VARIANT_ALLOW_READING_SHREDDED.key -> true.toString,
+          SQLConf.VARIANT_FORCE_SHREDDING_SCHEMA_FOR_TEST.key -> schema) {
+
+          df.write.format("delta").mode("append").saveAsTable("tbl")
+          // Make sure the actual parquet files are shredded
+          assert(numShreddedFiles(dir.getAbsolutePath, validation = { field: GroupType =>
+            field.getName == "v" && (field.getType("typed_value") match {
+              case t: GroupType =>
+                t.getFields.asScala.map(_.getName).toSet == Set("a", "b", "c")
+              case _ => false
+            })
+          }) == 1)
+          checkAnswer(
+            spark.read.format("delta").load(dir.getAbsolutePath).selectExpr("i", "to_json(v)"),
+            df.selectExpr("i", "to_json(v)").collect()
+          )
+        }
+      }
+    }
+  }
+
+  test("Set table property to invalid value") {
+    withTable("tbl") {
+      sql("CREATE TABLE tbl(s STRING, i INTEGER) USING DELTA")
+      val deltaLog = DeltaLog.forTable(spark, TableIdentifier("tbl"))
+      assert(!deltaLog.unsafeVolatileSnapshot.protocol
```

Comment:
Please use [DeltaLog.forTableWithSnapshot](https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/DeltaLog.scala#L872-L874) instead of the volatile snapshot:
```suggestion
      val (deltaLog, snapshot) = DeltaLog.forTableWithSnapshot(spark, TableIdentifier("tbl"))
      assert(!snapshot.protocol
```